presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java

```diff
@@ -313,6 +315,34 @@ protected Boolean visitInPredicate(InPredicate node, Void context)
         protected Boolean visitFunctionCall(FunctionCall node, Void context)
         {
             if (metadata.isAggregationFunction(node.getName())) {
+                if (node.getOrderBy().isPresent()) {
+                    List<Expression> sortKeys = node.getOrderBy().get().getSortItems().stream()
+                            .map(SortItem::getSortKey)
+                            .collect(toImmutableList());
+                    if (node.isDistinct()) {
+                        List<FieldId> argumentReferences = node.getArguments().stream()
+                                .map(NodeRef::of)
+                                .map(columnReferences::get)
+                                .collect(toImmutableList());
+                        for (Expression sortKey : sortKeys) {
+                            if (!argumentReferences.contains(columnReferences.get(NodeRef.of(sortKey)))) {
```

Comment:
Specifically, these tests are added in AbstractTestQueries:

        assertQuery(
                "SELECT x, array_agg(distinct z + y ORDER BY z + y DESC) FROM (VALUES (1, 2, 2), (2, 2, 3), (2, 4, 5), (3, 4, 4), (3, 2, 1), (1, 1, 1)) t(x, y, z) GROUP BY x",
                "VALUES (1, (4, 2)), (2, (9, 5)), (3, (8, 3))");

        assertQueryFails(
                "SELECT array_agg(DISTINCT x ORDER BY y) FROM (VALUES (1, 2), (3, 5), (4, 1)) t(x, y)",
                ".* For aggregate function with DISTINCT, ORDER BY expressions must appear in arguments");

        assertQueryFails(
                "SELECT array_agg(DISTINCT x+y ORDER BY y) FROM (VALUES (1, 2), (3, 5), (4, 1)) t(x, y)",
                ".* For aggregate function with DISTINCT, ORDER BY expressions must appear in arguments");

        assertQueryFails(
                "SELECT x, array_agg(distinct y ORDER BY z + y DESC) FROM (VALUES (1, 2, 2), (2, 2, 3), (2, 4, 5), (3, 4, 4), (3, 2, 1), (1, 1, 1)) t(x, y, z) GROUP BY x",
                ".* For aggregate function with DISTINCT, ORDER BY expressions must appear in arguments");